scripts/core/ui/components/popup.tsx

```diff
@@ -0,0 +1,128 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import {throttle, once} from 'lodash';
+import {createPopper, Instance as PopperInstance, Placement} from '@popperjs/core/';
+import maxSize from 'popper-max-size-modifier';
+import {applyMaxSize} from './AutoCompleteSuggestions';
+
+interface IPropsPositioner {
+    referenceElement: HTMLElement;
+    placement: Placement;
+    zIndex?: number;
+    onClose(): void;
+}
+
+class PopupPositioner extends React.PureComponent<IPropsPositioner> {
+    private wrapperEl: HTMLDivElement;
+    private positionOnce: (el: HTMLElement) => void;
+    private popper: PopperInstance;
+
+    constructor(props: IPropsPositioner) {
+        super(props);
+
+        this.closeOnClick = this.closeOnClick.bind(this);
+        this.closeOnScroll = throttle(this.closeOnScroll.bind(this), 200);
+
+        this.positionOnce = once((el) => {
+            this.popper = createPopper(
+                this.props.referenceElement,
+                el,
+                {
+                    placement: 'bottom',
+                    modifiers: [maxSize, applyMaxSize],
+                },
+            );
+        });
+    }
+
+    closeOnClick(event: MouseEvent) {
+        if (this.wrapperEl == null) {
+            return;
+        }
+
+        if (
+            this.props.referenceElement.contains(event.target as Node) !== true
+            && this.wrapperEl.contains(event.target as Node) !== true
+        ) {
+            this.props.onClose();
+        }
+    }
+
+    closeOnScroll(event: MouseEvent) {
+        if (this.wrapperEl == null) {
+            return;
+        }
+
+        if (this.wrapperEl.contains(event.target as Node) !== true) {
+            this.props.onClose();
+        }
+    }
+
+    componentDidMount() {
+        window.addEventListener('click', this.closeOnClick);
+        window.addEventListener('scroll', this.closeOnScroll, true);
+    }
+
+    componentWillUnmount() {
+        window.removeEventListener('click', this.closeOnClick);
+        window.removeEventListener('scroll', this.closeOnScroll, true);
+
+        this.popper.destroy();
+    }
+
+    render() {
+        return (
+            <div
+                ref={(el) => {
+                    this.wrapperEl = el;
+
+                    if (el != null) {
+                        if (el != null) {
+                            this.positionOnce(el);
```

Comment:
not sure why doing 2x `if (el != null)` and actually why not call this from componentDidMount?

Reply:
Just double checking to be sure :D It's interesting how I missed that. Regarding `componentDidMount`, I think `ref` function runs after `componentDidMount`, so I wouldn't have the element inside `componentDidMount`. I'm not 100% sure, so I'll double check that too ;D